### PR TITLE
Rund new runtime impl

### PR DESCRIPTION
**Pull request by quanweiZhou**
runtime implement

### CODE REVIEW

**Comment by zvonkok** [comment]
How does this PR compares to https://github.com/kata-containers/kata-containers/pull/2785?

File: src/runtime-rs/crates/shim/src/shim_run.rs [hunk 1]
```diff
@@ -36,12 +37,18 @@ impl ShimExecutor {
...
-            .context("new runtime server")?;
+        let mut service_manager = service::ServiceManager::new(
+            &self.args.id,
+            &self.args.publish_binary,
+            &self.args.address,
+            &self.args.namespace,
+            server_fd,
+        )
+        .await
+        .context("new runtime server")?;
```
Comment (liubin) [inline src/runtime-rs/crates/shim/src/shim_run.rs:51]: `new shim server` ?

File: src/runtime-rs/crates/shim/src/shim_delete.rs [hunk 2]
```diff
@@ -30,6 +31,16 @@ impl ShimExecutor {
...
         rsp.set_exited_at(exited_time);
 
+        let address = self
+            .socket_address(&self.args.id)
+            .context("socket address")?;
+        let trim_path = address.strip_prefix("unix://").context("trim path")?;
+        let file_path = Path::new("/").join(trim_path);
+        let file_path = file_path.as_path();
+        if std::fs::metadata(&file_path).is_ok() {
+            info!(sl!(), "cleanup the sandbox temp_dir: {:?}", &file_path);
```
Comment (liubin) [inline src/runtime-rs/crates/shim/src/shim_delete.rs:41]: Is the file a socket file or `sandbox temp_dir`?

File: src/runtime-rs/crates/service/src/manager.rs [hunk 3]
```diff
@@ -72,8 +152,13 @@ impl ServiceManager {
         Ok(())
     }
 
-    pub fn cleanup(id: &str) -> Result<()> {
-        RuntimeHandlerManager::cleanup(id)
+    pub fn cleanup(sid: &str) -> Result<()> {
+        let temp_dir = [KATA_PATH, sid].join("/");
+        if std::fs::metadata(temp_dir.as_str()).is_ok() {
+            info!(sl!(), "cleanup the sandbox temp_dir: {:?}", &temp_dir);
+            fs::remove_dir_all(temp_dir).ok();
```
Comment (liubin) [inline src/runtime-rs/crates/service/src/manager.rs:159]: What's `.ok()` used for?

File: src/runtime-rs/crates/service/src/manager.rs [hunk 4]
```diff
@@ -5,30 +5,91 @@
...
 /// message buffer size
 const MESSAGE_BUFFER_SIZE: usize = 8;
 
+pub const KATA_PATH: &str = "/run/kata";
+
 pub struct ServiceManager {
     receiver: Option<Receiver<Message>>,
     handler: Arc<RuntimeHandlerManager>,
     task_server: Option<Server>,
+    containerd_binary: String,
```
Comment (liubin) [inline src/runtime-rs/crates/service/src/manager.rs:39]: Note: also CRI-O should be considered.

File: src/runtime-rs/crates/runtimes/virt_container/src/container_manager/container.rs [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+            guest_hooks: None,
+            sandbox_pidns: false,
+            rootfs_mounts: vec![],
+        };
+
+        self.agent
+            .create_container(r)
+            .await
+            .context("agent create container")?;
+        self.resource_manager.dump().await;
```
Comment (liubin) [inline src/runtime-rs/crates/runtimes/virt_container/src/container_manager/container.rs:154]: Is this line for debugging?

File: src/runtime-rs/crates/runtimes/virt_container/src/container_manager/container.rs [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+        &self,
+        container_process: &ContainerProcess,
+    ) -> Result<ProcessWatcher> {
+        let logger = logger_with_process(container_process);
+        info!(logger, "start wait process");
+
+        let inner = self.inner.read().await;
+        inner
+            .fetch_exit_watcher(container_process)
+            .context("fetch exit watcher")
```
Comment (liubin) [inline src/runtime-rs/crates/runtimes/virt_container/src/container_manager/container.rs:247]: `exit` -> `exist` ?

File: src/runtime-rs/crates/runtimes/virt_container/src/container_manager/container.rs [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+    pub async fn stats(&self) -> Result<Option<agent::StatsContainerResponse>> {
+        let stats_resp = self
+            .agent
+            .stats_container(self.container_id.clone().into())
+            .await
+            .context("agent stats container")?;
+        Ok(Some(stats_resp))
+    }
+
+    // to do
```
Comment (liubin) [inline src/runtime-rs/crates/runtimes/virt_container/src/container_manager/container.rs:358]: `to do` what？

File: src/runtime-rs/crates/runtimes/virt_container/src/container_manager/container.rs [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+            resource.pids = None;
+            resource.block_io = None;
+            resource.hugepage_limits = Vec::new();
+            resource.network = None;
+        }
+
+        let mut ns: Vec<oci::LinuxNamespace> = Vec::new();
+        for n in linux.namespaces.iter() {
+            match n.r#type.as_str() {
+                oci::PIDNAMESPACE => continue,
```
Comment (liubin) [inline src/runtime-rs/crates/runtimes/virt_container/src/container_manager/container.rs:395]:
```rust
oci::PIDNAMESPACE | oci::NETWORKNAMESPACE => continue,
```

File: src/runtime-rs/crates/runtimes/virt_container/src/container_manager/container.rs [hunk 9]
```diff
@@ -0,0 +1,405 @@
...
+
+    pub async fn resize_pty(
+        &self,
+        process: &ContainerProcess,
+        width: u32,
+        height: u32,
+    ) -> Result<()> {
+        let logger = logger_with_process(process);
+        let inner = self.inner.read().await;
+        if inner.init_process.status == ProcessStatus::Running {
```
Comment (liubin) [inline src/runtime-rs/crates/runtimes/virt_container/src/container_manager/container.rs:334]: Copy miss?

File: src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+
+    async fn delete_process(&self, process: &ContainerProcess) -> Result<ProcessStateInfo> {
+        let container_id = &process.container_id.container_id;
+        match process.process_type {
+            ProcessType::Container => {
+                let mut containers = self.containers.write().await;
+                let c = containers
+                    .remove(container_id)
+                    .ok_or_else(|| Error::ContainerNotFound(container_id.to_string()))?;
+                c.state_process(process).await.context("state process")
```
Comment (liubin) [inline src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs:92]: What's the `state_process` meaning here?

File: src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+            None => {
+                warn!(logger, "failed to find watcher for wait process");
+            }
+        }
+
+        let status = status.read().await;
+
+        info!(logger, "wait process exit status {:?}", status);
+
+        // stop process
```
Comment (liubin) [inline src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs:177]: Question: why call `stop_process` here?

File: src/runtime-rs/crates/runtimes/virt_container/src/container_manager/process.rs [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+    pub width: u32,
+    pub status: ProcessStatus,
+
+    pub exit_status: Arc<RwLock<ProcessExitStatus>>,
+    pub exit_watcher_rx: Option<watch::Receiver<bool>>,
+    pub exit_watcher_tx: Option<watch::Sender<bool>>,
+    // used to sync between stdin io copy thread(tokio) and the close it call.
+    // close io call should wait until the stdin io copy finished to
+    // prevent stdin data lost.
+    pub wg_stdin: WaitGroup,
```
Comment (liubin) [inline src/runtime-rs/crates/runtimes/virt_container/src/container_manager/process.rs:50]: `WaitGroup` is not  a rust style variable name

File: src/runtime-rs/crates/runtimes/virt_container/src/health_check.rs [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+                    Err(mpsc::error::TryRecvError::Empty) => {
+                        // check agent
+                        match agent
+                            .check(agent::CheckRequest::new(""))
+                            .await
+                            .context("check health")
+                        {
+                            Ok(_) => {
+                                debug!(sl!(), "check {} agent health successfully", id);
+                                health_check_threshold_count += 1;
```
Comment (liubin) [inline src/runtime-rs/crates/runtimes/virt_container/src/health_check.rs:72]: Why check the version every 5 minutes?

File: src/runtime-rs/crates/runtimes/virt_container/src/sandbox.rs [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+
+#[async_trait]
+impl Sandbox for VirtSandbox {
+    async fn start(&self, netns: Option<String>, config: &TomlConfig) -> Result<()> {
+        let id = &self.sid;
+
+        // if sandbox running, return
+        // if sandbox not running try to start sandbox
+        let mut inner = self.inner.write().await;
+        if inner.state != SandboxState::Init {
```
Comment (liubin) [inline src/runtime-rs/crates/runtimes/virt_container/src/sandbox.rs:102]: This check conflicts with the comments above.

File: src/runtime-rs/crates/service/src/manager.rs (hunk 3 above)
Comment (Tim-Zhang) [reply src/runtime-rs/crates/service/src/manager.rs:159]: I think it is equal to `let _ = fs::remove_dir_all(temp_dir);` , the purpose is to ignore result.

File: src/libs/kata-sys-util/src/rand/mod.rs [hunk 15]
```diff
@@ -0,0 +1,10 @@
+// Copyright (c) 2019-2022 Alibaba Cloud
+// Copyright (c) 2019-2022 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+mod uuid;
+pub use uuid::UUID;
+mod random_bytes;
+pub use random_bytes::RandomBytes;
```
[the remark this answers is not in the document]
Comment (quanweiZhou) [reply src/libs/kata-sys-util/src/rand/mod.rs:8]: Thanks for your review, has been re-sorted.

File: src/runtime-rs/crates/shim/src/shim_run.rs (hunk 1 above)
Comment (quanweiZhou) [reply src/runtime-rs/crates/shim/src/shim_run.rs]: Thanks for your review, has been updated.

File: src/runtime-rs/crates/shim/src/shim_delete.rs (hunk 2 above)
Comment (quanweiZhou) [reply src/runtime-rs/crates/shim/src/shim_delete.rs]: Thanks for your review, it's a socket file, that has been updated.

File: src/runtime-rs/crates/service/src/manager.rs (hunk 3 above)
Comment (quanweiZhou) [reply src/runtime-rs/crates/service/src/manager.rs]: Thanks for your review, the purpose is to ignore the result, and add a comment.

File: src/runtime-rs/crates/service/src/manager.rs (hunk 4 above)
Comment (quanweiZhou) [reply src/runtime-rs/crates/service/src/manager.rs]: Thanks for your review, has been updated.

File: src/runtime-rs/crates/runtimes/virt_container/src/container_manager/container.rs (hunk 5 above)
Comment (quanweiZhou) [reply src/runtime-rs/crates/runtimes/virt_container/src/container_manager/container.rs:154]: Thanks for your review, it is trying to print resource info part of debugging, maybe remove this when supporting runtime metrics.

File: src/runtime-rs/crates/runtimes/virt_container/src/container_manager/container.rs (hunk 7 above)
Comment (quanweiZhou) [reply src/runtime-rs/crates/runtimes/virt_container/src/container_manager/container.rs]: Thanks for your review, has been removed.

File: src/runtime-rs/crates/runtimes/virt_container/src/container_manager/container.rs (hunk 8 above)
Comment (quanweiZhou) [reply src/runtime-rs/crates/runtimes/virt_container/src/container_manager/container.rs]: Thanks for your review, has been updated.

File: src/runtime-rs/crates/runtimes/virt_container/src/container_manager/container.rs (hunk 9 above)
Comment (quanweiZhou) [reply src/runtime-rs/crates/runtimes/virt_container/src/container_manager/container.rs]: Thanks for your review, it's trying to check process not running, change to `inner.init_process.status != ProcessStatus::Running`.

File: src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs (hunk 10 above)
Comment (quanweiZhou) [reply src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs:92]: Thanks for your review, `state_process` try to get process state for shim API need.

File: src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs (hunk 11 above)
Comment (quanweiZhou) [reply src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs:177]: Thanks for your review, this operation tries to clean up the container/process when the container/process exit.

File: src/runtime-rs/crates/runtimes/virt_container/src/health_check.rs (hunk 13 above)
Comment (quanweiZhou) [reply src/runtime-rs/crates/runtimes/virt_container/src/health_check.rs]: Thanks for your review, this is an issue for refactoring, the health check is in the 30s and the version check is 5min, has been updated.

File: src/runtime-rs/crates/runtimes/virt_container/src/sandbox.rs (hunk 14 above)
Comment (quanweiZhou) [reply src/runtime-rs/crates/runtimes/virt_container/src/sandbox.rs]: Thanks for your review, has been updated.

**Comment by quanweiZhou** [comment]
> How does this PR compares to #2785?

this PR is for runtime #2785 is for the agent

**Review by Tim-Zhang — APPROVED** [review]
LGTM, Thanks a lot @quanweiZhou

**Review by liubin — APPROVED** [review]

**Pull request merged**